specifyweb/frontend/js_src/lib/components/FormEditor/createView.ts

```diff
@@ -263,38 +263,68 @@ function createViewFromTemplate(
   const originalNames = filterArray(
     view.altViews.altViews.map(({ viewDef }) => viewDef)
   );
+
   const nameMapper = Object.fromEntries(
     originalNames.map((name) => [name, getUniqueDefinitionName(name, viewSets)])
   );
-  return {
-    ...viewSets,
-    views: [
-      ...viewSets.views,
-      {
-        ...view,
-        name,
-        altViews: {
-          ...view.altViews,
-          altViews: view.altViews.altViews.map((altView) => ({
-            ...altView,
-            viewDef:
-              typeof altView.viewDef === 'string'
-                ? nameMapper[altView.viewDef] ?? altView.viewDef
-                : altView.viewDef,
-          })),
-        },
-      },
-    ],
-    viewDefs: [
-      ...viewSets.viewDefs,
-      ...viewDefinitions.map((definition) => ({
-        ...definition,
+
+  const uniqueName = getUniqueDefinitionName(name, viewSets);
+
+  const updatedAltViews = view.altViews.altViews.map((altView) => ({
+    ...altView,
+    viewDef:
+      typeof altView.viewDef === 'string'
+        ? nameMapper[altView.viewDef] ?? altView.viewDef
+        : altView.viewDef,
+    name:
+      typeof altView.viewDef === 'string'
+        ? nameMapper[altView.viewDef] ?? altView.viewDef
+        : altView.viewDef!,
+  }));
+
+  const updatedView = {
+    ...view,
+    name: uniqueName,
+    altViews: { ...view.altViews, altViews: updatedAltViews },
+  };
+
+  const updatedViewDefinitions = viewDefinitions.map((definition) => ({
+    ...definition,
+    name: definition.name ? nameMapper[definition.name] : definition.name!,
+    raw: {
+      ...definition.raw,
+      attributes: {
+        ...definition.raw.attributes,
```

Comment:
was the bug caused by the fact that the `name` from line 298 was used instead of the name from line 293?
do you need to specify the `name` in both places or only one of the two?

note that `raw` key exists here as a way to defer parsing the viewdef (so that all viewdefs are not parsed when opening a view set file - viewdef is then only parsed when it's actually opened for editing)

see how raw XML string is gotten back from it here:
https://github.com/specify/specify7/blob/0cbeff429e84f653e83d0ef49ab5e9675d2c4660/specifyweb/frontend/js_src/lib/components/FormEditor/Editor.tsx#L198-L213
and see how attributes are cleared away in there (so that they are not editable in the UI)

but then see how the attributes are added back in here after user made edits to the view def itself:
https://github.com/specify/specify7/blob/0cbeff429e84f653e83d0ef49ab5e9675d2c4660/specifyweb/frontend/js_src/lib/components/FormEditor/Editor.tsx#L159-L170

I think the actual bug causing https://github.com/specify/specify7/issues/4134 might be that the raw.attributes.name is overriding the name
at the same time, I belive raw.attributes is not needed and should probably be set to empty object to avoid such issues (Because it is empty in all the other places in this file. e.x in https://github.com/specify/specify7/blob/0cbeff429e84f653e83d0ef49ab5e9675d2c4660/specifyweb/frontend/js_src/lib/components/FormEditor/createView.ts#L132-L141).

Thus, could you do the following changes:
1. set attributes to empty object here: https://github.com/specify/specify7/blob/0cbeff429e84f653e83d0ef49ab5e9675d2c4660/specifyweb/frontend/js_src/lib/components/FormEditor/spec.ts#L275
2. that would remove the need to set it to empty object here: https://github.com/specify/specify7/blob/0cbeff429e84f653e83d0ef49ab5e9675d2c4660/specifyweb/frontend/js_src/lib/components/FormEditor/Editor.tsx#L198-L213
3. and remove the need to set it back here: https://github.com/specify/specify7/blob/0cbeff429e84f653e83d0ef49ab5e9675d2c4660/specifyweb/frontend/js_src/lib/components/FormEditor/Editor.tsx#L159-L170
4. and remove the need to set raw.attributes in this file (createView.ts)

Reply:
because your first code snippet is not correct. To make attributes empty, you should have had something like this:
```ts
    raw: syncer<SimpleXmlNode, XmlNode>(
      (node) => ({
        ...defined(getOriginalSyncerInput(node), ''),
        attributes: {},
      }),
      toSimpleXmlNode
    ),
```

In your case, you first call defined() (which makes sure input is never undefined, and if it happens to be undefined, throw an error), and then use `??`, which only makes sense if you expect the value to the left to be undefined - but by calling `defined()` you ensure it's never undefined, so the code after `??` is never executed

can you try again with this change please?

(and if you get tired of this, I can take over when you want)

Reply:
I pushed a commit to this PR that implements the changes I suggested in https://github.com/specify/specify7/pull/4144/files#r1372511231 to simplify code a bit and prevent the bug with name being overridden from happening in the future.

In the process I also remembered why the contents of the `<definition>` tag wasn't replaced - the thing is that it shouldn't always be replaced. If `<definition>` is equal to `ObjectAttachment Form`, and your current view name is `CollectionObjectAttachment`, then even after cloning, the `<definition>` should still point to `ObjectAttachment Form`.

But if `<definition>` points to `CollectionObjectAttachment Form`, then it needs to be replaced with `CollectionObjectAttachment Form (2)` - so that is the 2nd bug in here - I also pushed a commit with the fix -please take a look and test the new behavior